File: rust/src/src/base/role/role_dao.rs
use color_eyre::eyre::Result;
use crate::common::context::{get_auth_model, Options};

use crate::r#gen::base::usr::usr_dao::find_by_id as find_by_id_usr;

// 角色
use crate::r#gen::base::role::role_dao::find_by_ids as find_by_ids_role;

use crate::r#gen::base::role::role_model::RoleModel;

/// 获取当前用户拥有的角色列表
#[allow(dead_code)]
pub async fn get_auth_role_models() -> Result<Vec<RoleModel>> {
  
  let aut_model = get_auth_model();
  if aut_model.is_none() {
    return Ok(vec![]);
  }
  
  let aut_model = aut_model.unwrap();
  
  let usr_id = aut_model.id;
  
  let options = Options::new()
    .set_is_debug(Some(false));
  let options = Some(options);
  
  let usr_model = find_by_id_usr(
    usr_id,
    options.clone(),
  ).await?;
  
  if usr_model.is_none() {
    return Ok(vec![]);
  }
  let usr_model = usr_model.unwrap();
  
  if usr_model.is_enabled == 0 {
    return Ok(vec![]);
  }
  
  let role_ids = usr_model.role_ids;
  
  let role_models = find_by_ids_role(
    role_ids,
    options,
  ).await?;
  
  Ok(role_models)
}

File: deno/src/base/role/role.dao.ts
import {
  getAuthModel,
} from "/lib/auth/auth.dao.ts";

import {
  findByIdUsr,
} from "/gen/base/usr/usr.dao.ts";

// 角色
import {
  findByIdsRole,
} from "/gen/base/role/role.dao.ts";

/**
 * 获取当前用户拥有的角色列表
 * 未登录用户返回角色空数组
 */
export async function getAuthRoleModels(): Promise<RoleModel[]> {
  const auth_model = await getAuthModel(false);
  if (!auth_model) {
    return [ ];
  }
  const usr_model = await findByIdUsr(
    auth_model.id,
    {
      is_debug: false,
    },
  );
  if (!usr_model || !usr_model.is_enabled) {
    return [ ];
  }
  const role_ids = usr_model.role_ids || [ ];
  const role_models = await findByIdsRole(
    role_ids,
    {
      is_debug: false,
    },
  );
  return role_models;
}


/**
 * 获取当前用户拥有的角色id列表
 */
export async function getAuthRoleIds() {
  const auth_model = await getAuthModel(false);
  if (!auth_model) {
    return [ ];
  }
  const usr_model = await findByIdUsr(
    auth_model.id,
    {
      is_debug: false,
    },
  );
  if (!usr_model || !usr_model.is_enabled) {
    return [ ];
  }
  const role_ids = usr_model.role_ids || [ ];
  return role_ids;
}

/**
 * 获取用户拥有的角色id列表
 * @param {UsrId} usr_id
 */
export async function getRoleIds(
  usr_id?: UsrId,
) {
  const usr_model = usr_id ? await findByIdUsr(
    usr_id,
    {
      is_debug: false,
    },
  ) : undefined;
  if (!usr_model || !usr_model.is_enabled) {
    return [ ];
  }
  const role_ids = usr_model.role_ids || [ ];
  return role_ids;
}
